### PR TITLE
A11-3900 Nri.Ui.ClickableText.V4

### PR DESCRIPTION
**Pull request by charbelrami**
# :wrench: Modifying a component

  - removes the `caption` attribute
  - removes `Css.display Css.inlineBlock`
  - inherits font family and weight from parent
  - adjusts font sizes
  - uses `Css.display Css.inlineFlex` only on buttons
  - removes `"-v2"` from data descriptor
  - uses dataDescriptor for `"clickable-text-label"`
  
Fixes:

- https://github.com/NoRedInk/noredink-ui/issues/1431
- https://github.com/NoRedInk/noredink-ui/issues/1364
- https://github.com/NoRedInk/noredink-ui/issues/1496

## Context

Fixes A11-3900

### Before

noredink-ui netlify app_(iPhone 14 Pro Max)

### After

localhost_8000_(iPhone 14 Pro Max)

## Component completion checklist

- [x] I've gone through the relevant sections of the [Development Accessibility guide](https://paper.dropbox.com/doc/Accessibility-guide-4-Development--BiIVdijSaoijjOuhz3iTCJJ1Ag-rGoHpC91pFg3zTrYpvOCQ) with the changes I made to this component in mind
- [x] Changes are clearly documented
  - [x] Component docs include a changelog
  - [x] Any new exposed functions or properties have docs
- [x] Changes extend to the Component Catalog
  - [x] The Component Catalog is updated to use the newest version, if appropriate
  - [x] The Component Catalog example version number is updated, if appropriate
  - [x] Any new customizations are available from the Component Catalog
  - [x] The component example still has:
    - an accurate preview
    - valid sample code
    - correct keyboard behavior
    - correct and comprehensive guidance around how to use the component
- [x] Changes to the component are tested/the new version of the component is tested
- [x] Component API follows standard patterns in noredink-ui
  - e.g., as a dev, I can conveniently add an `nriDescription`
  - and adding a new feature to the component will _not_ require major API changes to the component
- [x] If this is a new major version of the component, our team has stories created to upgrade all instances of the old component. Here are links to the stories:
  - add your story links here OR just write this is not a new major version
- [x] Please assign the following reviewers (applicable Sep 2023–Dec 2023):
  - [x] [a11y-volunteer-reviewers](https://github.com/orgs/NoRedInk/teams/volunteer-a11y-reviewers) - Someone from this group will review your PR in full.
  - [x]  [team-accessibilibats-a11ybats](https://github.com/orgs/NoRedInk/teams/team-accessibilibats-a11ybats) - Someone from this group will review your PR for ACCESSIBILITY ONLY.
  - [x]  Someone from your team who can review requirements from your team's perspective. (This could be the same person from the [a11y-volunteer-reviewers](https://github.com/orgs/NoRedInk/teams/volunteer-a11y-reviewers) group if you'd like.)
  - [x]  If there are user-facing changes, a designer. (You may want to direct your designer to the [deploy preview](https://github.com/NoRedInk/noredink-ui#reviews--preview-environments) for easy review):
    - For writing-related component changes, add Stacey (staceyadams)
    - For quiz engine-related components, add Ravi (ravi-morbia)
    - For a11y-related changes to general components, add Ben (bendansby)
    - For general component-related changes or if you’re not sure about something, add the Design group (NoRedInk/design)

### CODE REVIEW

**Comment by bendansby** [comment]
Hey Charbel, excited for this! I notice right off the bat that the inherited properties are causing some unintended side effects, like the links for each component in the CC:
image

I am wondering if we can restrict the inheritance to clickable text that is within a paragraph rather than have any clickable text inherit the properties of any parent?

**Comment by charbelrami** [comment]
@bendansby, I've made the changes!

**Comment by bendansby** [comment]
Thank you! 

Looking directly at the example, the inline links have a smaller font size than the parent paragraph:
image

**Review by bendansby — APPROVED** [review]
Looks great! Thanks for making these improvements!

**Review by tesk9 — COMMENTED** [review]
Code looks great! I really like the approach you took, with the new `Inherited` type constructor. Really clear and a great default behavior! Feels like a step toward rem replacing px 💪 

I noticed some alignment changes when there are icons that I want to flag (I will defer to Ben on whether anything is blocking or not):

| | Before | After | Description | 
| --- | --- | --- | --- |
| No right icon | Screenshot 2024-01-16 at 9 26 42 AM | Screenshot 2024-01-16 at 9 26 53 AM | The left icon now appears in line with the text when the text wraps, instead of staying centered relative to the total height of the text. |
| With right icon | Screenshot 2024-01-16 at 9 27 13 AM | Screenshot 2024-01-16 at 9 27 20 AM | The wrapping behavior of the right icon is now a little different depending on whether the clickable text is a button or link: if it's a button, it stays in line with the text instead of staying anchored on the right. |
| Docs and Source links | Screenshot 2024-01-16 at 9 30 52 AM | Screenshot 2024-01-16 at 9 30 58 AM | The Docs & Sources right icons look too high to me now? |
 
One other small thing -- could you please update the example page to say it's "V4"? Thanks!

**Comment by bendansby** [comment]
Thanks for that table! I think everything's OK except the right icon being too high. Was less obvious looking at the inline examples. Let's try to adjust that.

**Comment by bendansby** [comment]
One other thing I noticed: inline buttons with icons are now misaligned. Not a huge deal because this isn't very common, but I wonder if there's a way to give the icons a nominal height of zero or something so that they don't throw off the text?

Napkin 15 01-16-24, 8 55 08 AM

**Comment by krinhorn** [comment]
(Friendly reminder that reviewing the Percy output can help to catch some of these things! Happy to give you a tour of Percy and some tips if you'd like.)

**Review by bendansby — APPROVED** [review]
Awesome, I think it's ready. Thanks for all the iteration!

File: component-catalog/src/Examples/ClickableText.elm [hunk 1]
```diff
@@ -186,40 +181,48 @@ viewExamples ellieLinkConfig (State control) =
...
+            , ClickableText.icon UiIcon.starFilled
+            ]
+        , text ", "
+        , ClickableText.button "ClickableTexts with right icons"
+            [ ClickableText.appearsInline
+            , ClickableText.onClick (ShowItWorked moduleName "in-line button")
+            , ClickableText.rightIcon UiIcon.starFilled
+            ]
+        , text ", and "
+        , ClickableText.button "ClickableTexts with left and right icons"
```
Comment (tesk9) [inline component-catalog/src/Examples/ClickableText.elm:221]: Love this change! So clear 💯

File: src/Nri/Ui/ClickableText/V4.elm [hunk 2]
```diff
@@ -0,0 +1,721 @@
...
+            icon_
+                |> Svg.withCss
+                    ([ Css.width (Css.em 1)
+                     , Css.height (Css.em 1)
+                     , Css.position Css.relative
+                     , Css.top (Css.em 0.1)
+                     ]
+                        ++ (case position of
+                                Left ->
+                                    iconMarginRight
```
Comment (tesk9) [inline src/Nri/Ui/ClickableText/V4.elm:527]: Given the position absolute style, I don't think that the margin right is doing anything when the clickable is a button

File: src/Nri/Ui/ClickableText/V4.elm [hunk 3]
```diff
@@ -0,0 +1,721 @@
...
+                                                    , Css.left Css.zero
+                                                    ]
+
+                                                Link ->
+                                                    []
+                                           )
+                                        ++ config.iconStyles
+
+                                Right ->
+                                    iconMarginLeft
```
Comment (tesk9) [inline src/Nri/Ui/ClickableText/V4.elm:541]: Same thing here -- I don't think the iconMarginLeft does anything for buttons

**Review by tesk9 — APPROVED** [review]
Cool solution!

**Review by tesk9 — APPROVED** [review]
Fab, thank you!

**Pull request merged**